### DISCUSSION TITLE
Recovering from C-API appender flush failures

### DISCUSSION THREAD

**Discussion by chu-k** [duckdb/duckdb#15405] [Q&A]
I'm using the Appender to write large volumes of data that I'd like to be available to other reader connections simultaneously so I'm flushing quite frequently. 

(I apologize in advance, I'm not able to consistently repro this to set up a minimum example) . Questions first, and details below:
1. Is the initial flush error indicative of a bad write that isn't irrecoverable?
2. How can I abort the transaction that the appender appears to have opened, before recreating a new one?

Thanks in advance!

Occasionally, I run into the following error on flush. 
```
 Failed to create checkpoint because of error: Assertion triggered in file ".../duckdb/arm64-osx-dbg/src/execution/index/fixed_size_allocator.cpp\" on line 104: !mask.RowIsValid(offset)")
```

Which I try to recover by destroying and recreating a new appender on the same table:
```
duckdb_appender_destroy(&appender);

duckdb_appender new_appender;
if (duckdb_appender_create(conn_, NULL, table_name.c_str(), &new_appender) == DuckDBError) {
      auto err = std::string(duckdb_appender_error(appender));
```

This raises a transaction error: `Data Current transaction is aborted (please ROLLBACK)`. 

I destroy this new_appender also, then attempt to rollback explicitly with `duckdb_query(conn, 'ROLLBACK', NULL)`. Which then results in: `TransactionContext Error: cannot start a transaction within a transaction`

**Comment by Tishj** [comment]
That's an assertion failure, that's not supposed to happen and is likely a sign of a bug in the system
If you could create an issue with a reproduction for that we'd really appreciate that 👍

**Comment by chu-k** [comment]
Working on a repro, some more details in the meantime:
- The table that flush fails on has a UNIQUE composite key constraint, e.g.
```
CREATE TABLE Data (
    id INTEGER PRIMARY KEY,
    data_type TEXT,
    UNIQUE(id, data_type)
);
```
- Typically happens on subsequent opening of an existing db, WAL was at 15M in my case

**Comment by taniabogatsch** [comment]
Hi @chu-k - much has happened w.r.t. DuckDB code since you've opened this discussion. Is there any chance you managed to get a reproducer here?
